resources/js/tableofcontent.js

```diff
@@ -1,3 +1,75 @@
+/**
+ * Construct recurseData.str as a JSON Array that contains their header text, 
+ * link, and children headers.
+ * 
+ * References:
+ * https://stackoverflow.com/questions/187619/is-there-a-javascript-solution-to-generating-a-table-of-contents-for-a-page
+ * @param elem DOM element
+ * @param recurseData Object with fields: { level : int, toc : str, count : int, Set : levelSet }
+ */
+function recurseChild(elem, recurseData)
+{
+    if (elem !== "undefined") 
+    {
+        if(elem.nodeName.match(/^H\d+$/) && elem.textContent)
+        {
+            var headerText = elem.textContent.trim();
+            var prevLevel = recurseData.level;
+            var level = elem.nodeName.substr(1);
+            var anchor = "kiwix-toc-" + recurseData.count;
+            var anchorLink = window.location.href.replace(location.hash,"") + '#' + anchor;
+            recurseData.count += 1;
+
+            var anchorElem = document.createElement("a");
+            anchorElem.id = anchor;
+
+            /* Wrap header content with something we can reference. */
+            elem.insertAdjacentElement("afterbegin", anchorElem);
+
+            if (level < prevLevel)
+            {
+                /* Complete current element and the parent element.*/
+                recurseData.toc += ']}]}, ';
+            }
+            else if (level == prevLevel)
+            {
+                /* Complete current element*/
+                recurseData.toc += ']}, ';
+            }
+
+            recurseData.level = parseInt(level);
+            recurseData.levelSet.add(parseInt(level));
+            recurseData.toc += '{"text" : "' + headerText.replace(/"/g, '\\"') + '", "anchor": "' + anchorLink + '", ' + '"child" : [';
+        }
+
+        var c = elem.children;
+        for (var i = 0; i < c.length; i++)
+            recurseChild(c[i], recurseData);
+    }
+}
+
+function tocJSON()
+{
+    /* level used to track current header level.
+       toc used to store constructed list.
+       count used to uniquely identify each list item in toc.
+       levelSet used to retrieve the levels disregarding header level value.
+    */
+    var recurseData = { level: 0, toc: '{ "url" : "' + window.location.href.replace(location.hash,"") + '", "table" : [', count: 0, levelSet: new Set};
+    recurseChild(document.body, recurseData);
+
+    var levelArray = Array.from(recurseData.levelSet).sort();
+    levelArray.sort(function(a, b){return a - b});
+    var level = levelArray.indexOf(recurseData.level) + 1; 
+
+    /* End un-closed lists */
+    if (level)
+        recurseData.toc += (new Array(level + 1)).join(']}');
+    recurseData.toc += ']}';
+    return recurseData.toc;
+}
```

Comment:
I think that instead of constructing the final JSON string it is better to construct a recursive JS object and then call `JSON.stringify()` on it.

Reply:
Done